### PR TITLE
[ALLUXIO-3249] Channel#writeAndFlush get stuck in AbstractReadHandler

### PR DESCRIPTION
**Pull request by Reidddddd**
https://alluxio.atlassian.net/browse/ALLUXIO-3249

### CODE REVIEW

File: core/common/src/main/java/alluxio/network/netty/IdleReadWriteHandler.java [hunk 1]
```diff
@@ -19,22 +19,31 @@
...
+  public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
     if (evt instanceof IdleStateEvent) {
-      if (((IdleStateEvent) evt).state() == IdleState.WRITER_IDLE) {
+      IdleState state = ((IdleStateEvent) evt).state();
+      if (state == IdleState.WRITER_IDLE) {
         Protocol.Heartbeat heartbeat = Protocol.Heartbeat.newBuilder().build();
         ctx.writeAndFlush(new RPCProtoMessage(new ProtoMessage(heartbeat)));
+      } else if (state == IdleState.READER_IDLE) {
+        LOG.info("Netty reader is idle more than 'alluxio.network.netty.read.idle.timeout',"
+            + "closing context.");
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/network/netty/IdleReadWriteHandler.java]: `LOG.info("Netty reader is idle more than {}, closing context.", PropertyKey.Name. NETWORK_NETTY_READ_IDLE_TIMEOUT_MS)`

**Comment by Reidddddd** [comment]
yes, it will cause garbage collect this unhealthy channel.

**Comment by calvinjia** [comment]
@Reidddddd For the case when the channel is still healthy but not used (ie. we used it once and now have returned it to the pool), will this cause those channels to be closed?

We currently have something which automatically garbage collects unused channels (5 minutes), but that logic is only if it is returned to the pool (does not work for the case you are describing). I'm wondering for the normal case, will these two garbage collection mechanisms overlap?

**Comment by Reidddddd** [comment]
The phenomenon is that channel is open and is active (so judge is still healthy?), but server side writeAndFlush get stuck, and client side reads nothing, so the readLock is being held forever.

> I'm wondering for the normal case, will these two garbage collection mechanisms overlap?
I'm afraid it is, this pr may not be a good solution, but it does solve the described phenomenon. I'm not familiar with netty, if you come up a better solution, please point it out.

**Comment by Reidddddd** [comment]
BTW, the described phenomenon happens on spark client.

**Comment by calvinjia** [comment]
@Reidddddd Could you add a test which demonstrates the problem you are observing? Is this only a problem in `close`, in other calls I imagine the client will timeout in a reasonably short time when the server does not reply?

**Comment by calvinjia** [comment]
Jenkins, ok to test.

**Comment by Reidddddd** [comment]
I'm not sure it is easily reproducible in UT...

Would you mind spending some time to skim those pasted logs and descriptions in https://alluxio.atlassian.net/browse/ALLUXIO-3249?
It may help you know what's going on and enlighten you some thoughts.

**Comment by apc999** [comment]
@calvinjia do you have further suggestion on this PR?

**Comment by calvinjia** [comment]
@Reidddddd Sorry for the slow response. The problem with this change is we will implicitly set a 2 minute timeout for all Netty operations, which will be a problem for some calls like `close` which can take longer than 2 minutes due to uploading to object stores.

**Comment by Reidddddd** [comment]
Yep, please hold. I met some other issues caused by this fix in production env.

**Comment by ZacBlanco** [comment]
I don't think this code applies to the new RPC framework after Alluxio 2.0+.

There has also not been any activity for over a year. I'm going to close it.

**Pull request closed without merging**